crates/red_knot_python_semantic/src/types.rs

```diff
@@ -950,6 +950,15 @@ impl<'db> Type<'db> {
                     )
             }
 
+            // Any heterogeneous tuple type is assignable to `tuple`. This needs to be a special
+            // case because the left-hand side tuple might be a gradual type, so we can not rely
+            // on subtyping.
+            (Type::Tuple(_), Type::Instance(instance))
+                if instance.class.is_known(db, KnownClass::Tuple) =>
+            {
+                true
+            }
```

Comment:
I think we can generalise this: for any two types `T` and `X` where `T` is a heterogeneous tuple type and `X` is an arbitrary type, if `tuple` is assignable to `X` then `T` is also assignable to `X`, since any heterogeneous tuple type `T` will always be assignable to `tuple`, and the assignability relation is transitive

```suggestion
            (Type::Tuple(_), _) => KnownClass::Tuple.to_instance(db).is_assignable_to(target),
```